File: test/gammalasso.jl
# Comparing with Matt Taddy's gamlr.R
# To rebuild the test cases source(gammalasso.R)
using Lasso
using CSV, GLM, DataFrames, Random, SparseArrays

# often path length is different because of different stopping rules...
function issimilarhead(a::AbstractVector,b::AbstractVector;rtol=1e-4)
    n = min(size(a,1),size(b,1))
    isapprox(a[1:n],b[1:n];rtol=rtol) ? true : [a[1:n] b[1:n]]
end

function issimilarhead(a::AbstractMatrix,b::AbstractMatrix;rtol=1e-4)
    n = min(size(a,1),size(b,1))
    m = min(size(a,2),size(b,2))
    isapprox(a[1:n,1:m],b[1:n,1:m];rtol=rtol) ? true : [a[1:n,1:m] b[1:n,1:m]]
end

datapath = joinpath(dirname(@__FILE__), "data")

penaltyfactors = readcsvmat(joinpath(datapath,"penaltyfactors.csv"))

rtol=1e-2
Random.seed!(243214)
@testset "GammaLassoPath" begin
    @testset "$family" for (family, dist, link) in (("gaussian", Normal(), IdentityLink()), ("binomial", Binomial(), LogitLink()), ("poisson", Poisson(), LogLink()))
        data = readcsvmat(joinpath(datapath,"gamlr.$family.data.csv"))
        y = convert(Vector{Float64},data[:,1])
        X = convert(Matrix{Float64},data[:,2:end])
        (n,p) = size(X)
        @testset "penalty_factor=$pf" for pf = 1:size(penaltyfactors,2)
            penalty_factor = penaltyfactors[:,pf]
            if penalty_factor == ones(p)
                penalty_factor = nothing
            end
            @testset "γ=$γ" for γ in [0 2 10]
                fitname = "gamma$γ.pf$pf"

                # get gamlr.R prms and estimates
                prms = CSV.read(joinpath(datapath,"gamlr.$family.$fitname.params.csv"))
                fittable = CSV.read(joinpath(datapath,"gamlr.$family.$fitname.fit.csv"))
                gcoefs = readcsvmat(joinpath(datapath,"gamlr.$family.$fitname.coefs.csv");types=[Float64 for i=1:100])
                family = prms[1,Symbol("fit.family")]
                γ = prms[1,Symbol("fit.gamma")]
                λ = nothing #convert(Vector{Float64},fittable[Symbol("fit.lambda")]) # should be set to nothing evenatually

                # fit julia version
                glp = fit(GammaLassoPath, X, y, dist, link; γ=γ, stopearly=false,
                    λminratio=0.001, penalty_factor=penalty_factor, λ=λ,
                    standardize=false, standardizeω=false)

                # compare
                @test true==issimilarhead(glp.λ,fittable[Symbol("fit.lambda")];rtol=rtol)
                @test true==issimilarhead(glp.b0,fittable[Symbol("fit.alpha")];rtol=rtol)
                @test true==issimilarhead(convert(Matrix{Float64},glp.coefs'),gcoefs';rtol=rtol)
                # we follow GLM.jl convention where deviance is scaled by nobs, while in gamlr it is not
                @test true==issimilarhead(deviance(glp),fittable[Symbol("fit.deviance")]/nobs(glp);rtol=rtol)
                @test true==issimilarhead(deviance(glp,X,y),fittable[Symbol("fit.deviance")]/nobs(glp);rtol=rtol)
                # @test true==issimilarhead(round(df(glp)[2:end]),round(fittable[2:end,Symbol("fit.df")]))
                @test true==issimilarhead(loglikelihood(glp),fittable[Symbol("fit.logLik")];rtol=rtol)
                @test true==issimilarhead(aicc(glp),fittable[Symbol("fit.AICc")];rtol=rtol)

                # TODO: figure out why these are so off, maybe because most are corner solutions
                # and stopping rules for lambda are different
                # # what we really need all these stats for is that the AICc identifies the same minima:
                # if argmin(aicc(glp)) != lastindex(aicc(glp)) && argmin(fittable[Symbol("fit.AICc")]) != lastindex(fittable[Symbol("fit.AICc")])
                #     # interior minima
                #     println("comparing intereior AICc")
                #     @test argmin(aicc(glp)) == argmin(fittable[Symbol("fit.AICc")])
                # end

                # comparse CV, NOTE: this involves a random choice of train subsamples
                gcoefs_CVmin = vec(readcsvmat(joinpath(datapath,"gamlr.$family.$fitname.coefs.CVmin.csv")))
                gcoefs_CV1se = vec(readcsvmat(joinpath(datapath,"gamlr.$family.$fitname.coefs.CV1se.csv")))

                glp_CVmin = coef(glp,select=:CVmin,nCVfolds=10)
                glp_CV1se = coef(glp,select=:CV1se,nCVfolds=10)

                @test glp_CVmin ≈ gcoefs_CVmin rtol=0.3
                @test glp_CV1se ≈ gcoefs_CV1se rtol=0.3

                if γ==0
                    # Compare with LassoPath
                    lp = fit(LassoPath, X, y, dist, link; stopearly=false,
                        λminratio=0.001, penalty_factor=penalty_factor, λ=λ,
                        standardize=false, standardizeω=false)
                    @test glp.λ == lp.λ
                    @test glp.b0 ≈ lp.b0
                    @test glp.coefs ≈ lp.coefs
                end
            end
        end
    end
end
# ## the following code is useful for understanding comparison failures
#
# rdist(x::Number,y::Number) = abs(x-y)/max(abs(x),abs(y))
# rdist(x::AbstractArray{T}, y::AbstractArray{S}; norm::Function=norm) where {T<:Number,S<:Number} = norm(x - y) / max(norm(x), norm(y))

File: test/gammalasso.R
# creates test data for gammalasso.jl
library(gamlr)

setwd("~/.julia/dev/Lasso/test/data")

### a low-D test (highly multi-collinear)

set.seed(13)
n <- 1000
p <- 3
standardize = FALSE
xvar <- matrix(0.9, nrow=p,ncol=p)
diag(xvar) <- 1
x <- matrix(rnorm(p*n), nrow=n)%*%chol(xvar)
Ey <- 4 + 3*x[,1] + -1*x[,2]
y <- Ey + rnorm(n)

rbinom1 <- function(n,p) {rbinom(n,1,p)}
pop.var <- function(x) var(x) * (length(x)-1) / length(x)
pop.sd <- function(x) sqrt(pop.var(x))

families <- list(gaussian(link="identity"),binomial(link="logit"),poisson(link="log"))
randfuns <- list(rnorm,rbinom1,rpois)
gammas <- c(0, 2, 10)
penaltyfactors <- list(c(1, 1, 1), c(0, 1, 0.4), c(0.01, 1, 0.4))

write.table(penaltyfactors,file = "penaltyfactors.csv", sep=",",row.names=FALSE,col.names=FALSE)

for(f in 1:length(families)) {

  family <- families[[f]]
  print(family)
  randfun <- randfuns[[f]]
  invlink <- family$linkinv
  yp <- randfun(n,invlink(Ey))
  if (family$family == "gaussian") {
    print("standardizing y for comparability with glmnet")
    yp <- yp / pop.sd(yp)
  }

  # export test data
  data = cbind(yp, x)
  familyfilename <- paste0("gamlr.",family$family)
  write.table(data,file = paste0(familyfilename,".data.csv"), sep=",",row.names=FALSE,col.names=FALSE)

  for(pf in 1:length(penaltyfactors)) {
    penaltyfactor <- penaltyfactors[[pf]]

    # export estimates
    for(gamma in gammas) {
      # AICc selection
      fit <- gamlr(x, yp, gamma=gamma, lambda.min.ratio=1e-3, family=family$family, varweight=penaltyfactor, standardize=standardize)
      fitname <- paste0("gamma",gamma,".pf",pf)
      fitfilename <- paste0(familyfilename,".",fitname)

      fit$AICc <- AICc(fit)
      fit$logLik <- logLik(fit)
      fittable <- data.frame(fit$lambda,fit$df,fit$deviance,fit$alpha,fit$logLik,fit$AICc)
      write.table(fittable,file = paste0(fitfilename,".fit.csv") ,sep=",",row.names=FALSE)

      coefs <- as.matrix(fit$beta)
      write.table(coefs,file = paste0(fitfilename,".coefs.csv") ,sep=",",row.names=FALSE,col.names=FALSE)

      # now 10-fold cv
      cvfit <- cv.gamlr(x, yp, gamma=gamma, lambda.min.ratio=1e-3, family=family$family, standardize=standardize, nfold=10)

      coefs_cvmin <- as.matrix(coef(cvfit,select="min"))
      write.table(coefs_cvmin,file = paste0(fitfilename,".coefs.CVmin.csv") ,sep=",",row.names=FALSE,col.names=FALSE)

      coefs_cv1se <- as.matrix(coef(cvfit,select="1se"))
      write.table(coefs_cv1se,file = paste0(fitfilename,".coefs.CV1se.csv") ,sep=",",row.names=FALSE,col.names=FALSE)

      # both fit's params
      params <- data.frame(fit$gamma,fit$family,cvfit$seg.min,cvfit$seg.1se,cvfit$lambda.min,cvfit$lambda.1se)
      write.table(params,file = paste0(fitfilename,".params.csv") ,sep=",",row.names=FALSE)

      }
  }
}
